### PR TITLE
[stable/prometheus] v3.0.0

### PR DESCRIPTION
**Pull request by mgoodness**
In which I pull in our current best practices and do some reorg. End goal is increased configurability, flexibility, usability, and reliability.

TODO:
- [x] Provide node exporter DaemonSet option

### CODE REVIEW

**Comment by mgoodness** [comment]
cc: everyone who's opened an issue or PR on Prometheus - @danielfm @ravishivt @vishh @mstrzele @ged15 @lchi @unguiculus

**Comment by danielfm** [comment]
I didn't test it, but looks good to me! Just make it explicit in README that `extraArgs` are dicts! :P

What I don't understand though is why it's necessary to provide a StatefulSet option for Prometheus/AlertManager. What is there to gain by using StatefulSets instead of Deployments?

**Comment by mgoodness** [comment]
@danielfm I can't believe I missed that. Fixed :)

I haven't actually decided on providing Prometheus as a StatefulSet. AFAIK, it's not meant to be clustered, so it may not make sense. Thoughts & opinions are welcome.

Alertmanager, OTOH, does support clustered configuration for HA. Biggest argument in its favor is that if a node running Alertmanager goes down, it might be nice to be alerted!

**Comment by danielfm** [comment]
@mgoodness 

> I haven't actually decided on providing Prometheus as a StatefulSet. AFAIK, it's not meant to be clustered, so it may not make sense. Thoughts & opinions are welcome.

Perhaps the one thing I can think of right now that favors StatefulSets for Prometheus is that deleting and/or scaling a StatefulSet down [will not delete the volumes associated with it](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#limitations), so maybe using StatefulSets is a good idea after all.

> Alertmanager, OTOH, does support clustered configuration for HA. Biggest argument in its favor is that if a node running Alertmanager goes down, it might be nice to be alerted!

Yes, that makes sense!

**Comment by unguiculus** [comment]
I also think StatefulSets make sense. BTW, the Prometheus Operator uses StatefulSets as well.

What's the future of this chart anyways? I'm following your chart solution for the operator (https://github.com/mgoodness/kube-prometheus-chart). Will we be able to switch to it any time soon?

**Comment by mgoodness** [comment]
@unguiculus The Operator's use of StatefulSet for Prometheus surprised me. I'm still not exactly sure what's gained over a Deployment/ReplicaSet. But I've not done a ton of research, either.

My goal is to make the Operator chart (with kube-prometheus config) the standard for deploying Prometheus on Kubernetes. It'll be so much easier to setup and maintain using K8s-native mechanisms. The problem is that ThirdPartyResources really aren't ready for prime time yet. Helm can't create them, for example, because `kubectl apply` doesn't work as expected. The best I can do right now is use NOTES.txt to output instructions. Forget about upgrading, and even uninstalling is a crapshoot.

All that said, I'm definitely going to keep working on that chart as the upstream projects progress. In the meantime, I want to make this chart as easy to use as possible.

**Comment by unguiculus** [comment]
Helm 2.2.0 seems to fix TPR issues. Would that allow you to proceed with the Prometheus Operator? 

https://github.com/kubernetes/helm/releases/tag/v2.2.0

**Comment by mgoodness** [comment]
@unguiculus I'll be investigating that this week!

All: with Helm 2.2.0 now supporting [conditional dependencies](https://github.com/kubernetes/helm/blob/master/docs/charts.md#tags-and-condition-fields-in-requirementsyaml), I'm thinking about pulling alertmanager, kube-state-metrics, and node exporter out into their own charts. In fact, @fiunchinho has already submitted a stand-alone kube-state-metrics chart in #662. The Prometheus chart would install these other components as requested. Thoughts?

**Comment by mgoodness** [comment]
Quick follow-up to my last comment. Due to an impending job change, and after discussion with the other maintainers, I've decided not to componentize yet. Instead, I'll work on providing the StatefulSet options, then get this PR merged as a monolith.

Pulling components out into individual charts and using the new dependency system will be my focus for the next version.

**Comment by mgoodness** [comment]
I'm abandoning my effort to add StatefulSet options in this version. Deepest apologies to anyone who was counting on that functionality; I simply don't have any more cycles to spend at this time.

Consider this ready for review.

**Comment by mgoodness** [comment]
@k8s-bot ok to test

**Comment by cemo** [comment]
😭

**Comment by mgoodness** [comment]
Has anyone else deployed this version? Would appreciate some third-party reviews so it can be merged.

**Comment by danielfm** [comment]
@mgoodness not yet, I'm on vacation right now... I'll upgrade to this version as soon as I get back (in a week or so).

File: stable/prometheus/Chart.yaml [hunk 1]
```diff
@@ -1,10 +1,12 @@
 name: prometheus
-version: 2.0.4
-description: A Prometheus Helm chart for Kubernetes. Prometheus is a monitoring system and time series database.
+version: 3.0.0
+description: Prometheus is a monitoring system and time series database.
 home: https://prometheus.io/
+icon: https://github.com/prometheus/prometheus.github.io/blob/master/assets/prometheus_logo-cb55bb5c346.png
```
Comment (prydonius) [inline stable/prometheus/Chart.yaml]: For this to work in Monocular it needs the raw link, e.g. https://raw.githubusercontent.com/prometheus/prometheus.github.io/master/assets/prometheus_logo-cb55bb5c346.png

**Comment by prydonius** [comment]
Took a quick pass, code review lgtm - I'll delegate UX testing to someone who knows more about Prometheus!

File: stable/prometheus/templates/NOTES.txt [hunk 2]
```diff
@@ -45,18 +47,19 @@ http://{{ . }}
...
-{{- else if contains "LoadBalancer" .Values.alertmanager.serviceType }}
+{{- else if contains "LoadBalancer" .Values.alertmanager.service.type }}
   NOTE: It may take a few minutes for the LoadBalancer IP to be available.
         You can watch the status of by running 'kubectl get svc --namespace {{ .Release.Namespace }} -w {{ template "alertmanager.fullname" . }}'
+
   export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "alertmanager.fullname" . }} -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
-  echo http://$SERVICE_IP:{{ .Values.alertmanager.httpPort }}
-{{- else if contains "ClusterIP"  .Values.alertmanager.serviceType }}
-  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "fullname" . }},component={{ .Values.alertmanager.name }}" -o jsonpath="{.items[0].metadata.name}")
+  echo http://$SERVICE_IP:{{ .Values.alertmanager.service.servicePort }}
```
Comment (sstarcher) [inline stable/prometheus/templates/NOTES.txt:55]: output on the notes is difficult to read.  Maybe some slight changes in newlines and prefixing with AlertManager / Prometheus Server on the lines to segment

```
NOTES:
The Prometheus server can be accessed via port 80 on the following DNS name from within your cluster:
prometheus-prometheus-server.infra.svc.cluster.local


Get the server URL by running these commands in the same shell:
  export POD_NAME=$(kubectl get pods --namespace infra -l "app=prometheus,component=server" -o jsonpath="{.items[0].metadata.name}")
  kubectl --namespace infra port-forward $POD_NAME 9090
The Prometheus alertmanager can be accessed via port 80 on the following DNS name from within your cluster:
prometheus-prometheus-alertmanager.infra.svc.cluster.local


Get the server URL by running these commands in the same shell:
  export POD_NAME=$(kubectl get pods --namespace infra -l "app=prometheus,component=alertmanager" -o jsonpath="{.items[0].metadata.name}")
  kubectl --namespace infra port-forward $POD_NAME 9093

For more information on running Prometheus, visit:
https://prometheus.io/
```

File: stable/prometheus/templates/alertmanager-configmap.yaml [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
+    app: {{ template "name" . }}
+    chart: {{ .Chart.Name }}-{{ .Chart.Version }}
     component: "{{ .Values.alertmanager.name }}"
-    heritage: "{{ .Release.Service }}"
-    release: "{{ .Release.Name }}"
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
   name: {{ template "alertmanager.fullname" . }}
 data:
 {{ toYaml .Values.alertmanagerFiles | indent 2 }}
```
Comment (sstarcher) [inline stable/prometheus/templates/alertmanager-configmap.yaml:13]: Should this be documented in the README's table?  It's in the readme, but not in the table.

Comment (sstarcher) [reply stable/prometheus/templates/alertmanager-configmap.yaml:13]: Would it make sense to include some helpful defaults https://github.com/kayrus/prometheus-kubernetes/tree/master/prometheus-rules

Comment (sstarcher) [reply stable/prometheus/templates/alertmanager-configmap.yaml:13]: hmm my mistake those values would go under serverFiles.rules

**Comment by mgoodness** [comment]
Improved (I think) the `NOTES.txt` output. Also added `alertmanagerFiles` & `serverFiles` to the README table. @sstarcher, wanna give it another look?

**Comment by mikesplain** [comment]
@mgoodness This looks awesome. I deployed locally and pushed to a test cluster a little bit ago.  The built in node exporter makes this chart much more useful.  The only change I had to make was decrease `scrape_interval` for prometheus (defaults to 1m, had to set to 30s) to make some of the existing grafana dashboards like (https://grafana.com/dashboards/1621) work. That said, since you can pass the entire CM as a value, that's not a real problem. Great job, hope to see this merged in soon!

**Comment by sstarcher** [comment]
I know this is essentially the practice that most chart developers follow, but I find it very painful.

When a release name is set to the name of the chart every resource tends to be stuttered.  
If I do `helm install --name=prometheus stable/prometheus`. I get service that look like `prometheus-prometheus-server`

File: stable/prometheus/README.md [hunk 4]
```diff
@@ -42,69 +42,120 @@ The command removes all the Kubernetes components associated with the chart and
...
+`alertmanagerFiles` | alertmanager ConfigMap entries | `alertmanager.yml`
+`configmapReload.name` | configmap-reload container name | `configmap-reload`
+`configmapReload.image.repository` | configmap-reload container image repository | `jimmidyson/configmap-reload`
+`configmapReload.image.tag` | configmap-reload container image tag | `v0.1`
+`configmapReload.image.pullPolicy` | configmap-reload container image pull policy | `IfNotPresent`
+`configmapReload.resources` | configmap-reload pod resource requests & limits | `{}`
+`kubeStateMetrics.enabled` | If true, create kube-state-metrics | `true`
+`kubeStateMetrics.name` | kube-state-metrics container name | `kube-state-metrics`
+`kubeStateMetrics.image.repository` | kube-state-metrics container image repository| `gcr.io/google_containers/kube-state-metrics`
+`kubeStateMetrics.image.tag` | kube-state-metrics container image tag | `v0.3.0`
```
Comment (sstarcher) [inline stable/prometheus/README.md]: the default is   tag: v0.4.1

**Comment by mgoodness** [comment]
> I know this is essentially the practice that most chart developers follow, but I find it very painful.

> When a release name is set to the name of the chart every resource tends to be stuttered.
If I do helm install --name=prometheus stable/prometheus. I get service that look like prometheus-prometheus-server

I understand. However, we need to allow multiple installations of a chart into the same namespace. Using the release name as a prefix is the most convenient way to create unique resource names.

**Comment by sstarcher** [comment]
@mgoodness I certainly agree the resources should be unique, but it would be nice if Release.Name == Chart.Name that it would not repeat.  As the Release.Name always has to be unique.

**Comment by mgoodness** [comment]
@sstarcher I'm a big fan of consistency and would rather not see all sorts of custom naming patterns and exceptions among charts. However, if there's community consensus we can certainly add it later. I don't consider it a blocker, though.

File: stable/prometheus/values.yaml [hunk 5]
```diff
@@ -307,176 +413,190 @@ serverFiles:
...
+    # `labelmap` relabeling action.
+
+    # Scrape config for API servers.
+    #
+    # Kubernetes exposes API servers as endpoints to the default/kubernetes
+    # service so this uses `endpoints` role and uses relabelling to only keep
+    # the endpoints associated with the default/kubernetes service using the
+    # default named port `https`. This works for single API server deployments as
+    # well as HA API server deployments.
+    scrape_configs:
```
Comment (mikesplain) [inline stable/prometheus/values.yaml]: Declaring `scrape_configs` again here overrides the static prometheus config above. Should probably be removed.

**Comment by sstarcher** [comment]
node_exporter instances label contains the port at the end.  This should probably be relabeled to just contain the IP.  

Example: `instance: 172.20.66.160:9100`

**Comment by mgoodness** [comment]
> node_exporter instances label contains the port at the end. This should probably be relabeled to just contain the IP.

We're really only providing Prometheus configuration as a working default, copied (more or less) straight from the upstream project. Further customization is left to the end-user.

**Comment by cemo** [comment]
@mgoodness thanks for your hard work. I have a little questions and did not want to open an issue. I am using some dashboards of prometheus operator and seen metrics like `container_network_receive_bytes_total` and `container_cpu_usage_seconds_total`.

1. Do you have any idea where these metrics are coming from? (seems kube-proxy based on my researches)
2. If this is coming from kube-proxy, how can I add necessary scraping configuration?

**Comment by andrewhowdencom** [comment]
@cemo I think they come from the kubelet (or heapster)

**Comment by dkoshkin** [comment]
@cemo did you ever figure out the graphs with `container_network_receive_bytes_total` and `container_cpu_usage_seconds_total` not working, I can see them in the prometheus dashboard but it shows up as "no data points" in Grafana.

**Comment by cemo** [comment]
@dkoshkin Yes, It is because of default scrape interval. :) Prometheus rate function requires at least 2 vector however there is only one single vector for 1m interval which is equal to scrape interval. Try to change `rate` function to 2m. Let me know in case a problem.

**Pull request merged**